## Data Processing

In [1]:
import pandas as pd

df_ff6 = pd.read_csv('/content/FF6.csv')
df_cay = pd.read_csv('/content/cay_index.csv')
df_market = pd.read_csv('/content/Market-Returns-Cleaned.csv')
df_consumption = pd.read_csv('/content/Consumption-Cleaned.csv')
df_ff6_factors = pd.read_csv('/content/FF6_Factors.csv')
df_income = pd.read_csv('/content/Income-Cleaned.csv')
df_risk_free = pd.read_csv('/content/Risk-Free-Cleaned.csv')

print("FF6.csv:")
display(df_ff6.head())
df_ff6.info()

print("\ncay_index.csv:")
display(df_cay.head())
df_cay.info()

print("\nMarket-Returns-Cleaned.csv:")
display(df_market.head())
df_market.info()

print("\nConsumption-Cleaned.csv:")
display(df_consumption.head())
df_consumption.info()

print("\nFF6_Factors.csv:")
display(df_ff6_factors.head())
df_ff6_factors.info()

print("\nIncome-Cleaned.csv:")
display(df_income.head())
df_income.info()

print("\nRisk-Free-Cleaned.csv:")
display(df_risk_free.head())
df_risk_free.info()

FF6.csv:


,month,Portfolios BV returns (%),Portfolios BN returns (%),Portfolios BG returns (%),Portfolios SV returns (%),Portfolios SN returns (%),Portfolios SG returns (%)
0,Apr-14,NaN,0.262471,0.172784,13.215249,6.550439,5.111012
1,May-14,NaN,23.102455,7.275463,27.552673,24.288677,15.810411
2,Jun-14,NaN,1.873951,7.745938,18.652457,16.839867,13.023586
3,Jul-14,NaN,-3.014671,1.413907,-7.500655,-2.044578,2.165995
4,Aug-14,NaN,-2.396891,4.203748,-0.729519,4.765247,4.484203


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132 entries, 0 to 131
Data columns (total 7 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   month                      132 non-null    object 
 1   Portfolios BV returns (%)  115 non-null    float64
 2   Portfolios BN returns (%)  132 non-null    float64
 3   Portfolios BG returns (%)  132 non-null    float64
 4   Portfolios SV returns (%)  132 non-null    float64
 5   Portfolios SN returns (%)  132 non-null    float64
 6   Portfolios SG returns (%)  132 non-null    float64
dtypes: float64(6), object(1)
memory usage: 7.3+ KB

cay_index.csv:


,month_x,cay
0,2014-04,0.057466
1,2014-05,-0.025694
2,2014-06,0.027865
3,2014-07,0.094101
4,2014-08,0.074530


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132 entries, 0 to 131
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   month_x  132 non-null    object 
 1   cay      132 non-null    float64
dtypes: float64(1), object(1)
memory usage: 2.2+ KB

Market-Returns-Cleaned.csv:


,month,market_return,Unnamed: 2,Unnamed: 3,Unnamed: 4,Unnamed: 5,Unnamed: 6,Unnamed: 7,Unnamed: 8,Unnamed: 9,...,Unnamed: 113,Unnamed: 114,Unnamed: 115,Unnamed: 116,Unnamed: 117,Unnamed: 118,Unnamed: 119,Unnamed: 120,Unnamed: 121,Unnamed: 122
0,Apr-14,-0.006342,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,May-14,0.008682,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,Jun-14,0.015540,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,Jul-14,-0.004266,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,Aug-14,0.001078,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132 entries, 0 to 131
Columns: 123 entries, month  to Unnamed: 122
dtypes: float64(122), object(1)
memory usage: 127.0+ KB

Consumption-Cleaned.csv:


,month,weighted_avg_consumption
0,2014-04,7208.683747
1,2014-05,7005.035044
2,2014-06,7311.807185
3,2014-07,8051.858683
4,2014-08,7449.553517


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132 entries, 0 to 131
Data columns (total 2 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   month                     132 non-null    object 
 1   weighted_avg_consumption  132 non-null    float64
dtypes: float64(1), object(1)
memory usage: 2.2+ KB

FF6_Factors.csv:


,Date,SMB,HML,WML,MF,RF
0,2014-04,5.592209,7.810347,-3.415147,0.013121,0.701553
1,2014-05,4.340620,10.499133,-8.697000,9.950842,0.693805
2,2014-06,9.644133,5.182859,2.562922,6.929204,0.699993
3,2014-07,0.896284,-9.319494,-1.127898,-0.233564,0.700633
4,2014-08,3.732569,-4.896134,5.494075,2.452235,0.637590


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132 entries, 0 to 131
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Date    132 non-null    object 
 1   SMB     132 non-null    float64
 2   HML     132 non-null    float64
 3   WML     132 non-null    float64
 4   MF      132 non-null    float64
 5   RF      132 non-null    float64
dtypes: float64(5), object(1)
memory usage: 6.3+ KB

Income-Cleaned.csv:


,month,national_weighted_avg_income
0,2014-04,12811.15146
1,2014-05,14071.83162
2,2014-06,13803.13302
3,2014-07,14442.45259
4,2014-08,13169.36297


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132 entries, 0 to 131
Data columns (total 2 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   month                         132 non-null    object 
 1   national_weighted_avg_income  132 non-null    float64
dtypes: float64(1), object(1)
memory usage: 2.2+ KB

Risk-Free-Cleaned.csv:


,month,Yeild
0,2014-04,8.8550
1,2014-05,8.6456
2,2014-06,8.5619
3,2014-07,8.6456
4,2014-08,8.6038


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132 entries, 0 to 131
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   month   132 non-null    object 
 1   Yeild   132 non-null    float64
dtypes: float64(1), object(1)
memory usage: 2.2+ KB


In [2]:
df_ff6['month'] = pd.to_datetime(df_ff6['month'], format='%b-%y')
df_ff6_factors['Date'] = pd.to_datetime(df_ff6_factors['Date'])
df_market['month '] = pd.to_datetime(df_market['month '], format='%b-%y')
df_risk_free['month'] = pd.to_datetime(df_risk_free['month'])
df_cay['month_x'] = pd.to_datetime(df_cay['month_x'])
df_consumption['month'] = pd.to_datetime(df_consumption['month'])
df_income['month'] = pd.to_datetime(df_income['month'])

merged_df = pd.merge(df_ff6, df_risk_free, left_on='month', right_on='month', how='inner')
merged_df = pd.merge(merged_df, df_ff6_factors, left_on='month', right_on='Date', how='inner')
merged_df = pd.merge(merged_df, df_market[['month ', 'market_return']], left_on='month', right_on='month ', how='inner')

portfolio_cols = ['Portfolios BV returns (%)', 'Portfolios BN returns (%)', 'Portfolios BG returns (%)', 'Portfolios SV returns (%)', 'Portfolios SN returns (%)', 'Portfolios SG returns (%)']
for col in portfolio_cols:
    merged_df[f'{col}_excess'] = merged_df[col] - merged_df['Yeild']

merged_df['market_return_excess'] = merged_df['market_return'] - merged_df['Yeild']

merged_df.rename(columns={'Date': 'date', 'month_x': 'date', 'month': 'date', 'month ': 'date',
                          'Portfolios BV returns (%)_excess': 'BV_excess_return',
                          'Portfolios BN returns (%)_excess': 'BN_excess_return',
                          'Portfolios BG returns (%)_excess': 'BG_excess_return',
                          'Portfolios SV returns (%)_excess': 'SV_excess_return',
                          'Portfolios SN returns (%)_excess': 'SN_excess_return',
                          'Portfolios SG returns (%)_excess': 'SG_excess_return',
                          'market_excess_return': 'market_excess_return',
                          'SMB': 'smb', 'HML': 'hml', 'WML': 'wml', 'MF': 'mf', 'RF': 'rf',
                          'weighted_avg_consumption': 'consumption',
                          'national_weighted_avg_income': 'income',
                          'cay': 'cay_index'}, inplace=True)


columns_to_drop = portfolio_cols + ['market_return', 'Yeild']
merged_df.drop(columns=columns_to_drop, inplace=True)

display(merged_df.head())
merged_df.info()

,date,date,smb,hml,wml,mf,rf,date,BV_excess_return,BN_excess_return,BG_excess_return,SV_excess_return,SN_excess_return,SG_excess_return,market_return_excess
0,2014-04-01,2014-04-01,5.592209,7.810347,-3.415147,0.013121,0.701553,2014-04-01,NaN,-8.592529,-8.682216,4.360249,-2.304561,-3.743988,-8.861342
1,2014-05-01,2014-05-01,4.340620,10.499133,-8.697000,9.950842,0.693805,2014-05-01,NaN,14.456855,-1.370137,18.907073,15.643077,7.164811,-8.636918
2,2014-06-01,2014-06-01,9.644133,5.182859,2.562922,6.929204,0.699993,2014-06-01,NaN,-6.687949,-0.815962,10.090557,8.277967,4.461686,-8.546360
3,2014-07-01,2014-07-01,0.896284,-9.319494,-1.127898,-0.233564,0.700633,2014-07-01,NaN,-11.660271,-7.231693,-16.146255,-10.690178,-6.479605,-8.649866
4,2014-08-01,2014-08-01,3.732569,-4.896134,5.494075,2.452235,0.637590,2014-08-01,NaN,-11.000691,-4.400052,-9.333319,-3.838553,-4.119597,-8.602722


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132 entries, 0 to 131
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   date                  132 non-null    datetime64[ns]
 1   date                  132 non-null    datetime64[ns]
 2   smb                   132 non-null    float64       
 3   hml                   132 non-null    float64       
 4   wml                   132 non-null    float64       
 5   mf                    132 non-null    float64       
 6   rf                    132 non-null    float64       
 7   date                  132 non-null    datetime64[ns]
 8   BV_excess_return      115 non-null    float64       
 9   BN_excess_return      132 non-null    float64       
 10  BG_excess_return      132 non-null    float64       
 11  SV_excess_return      132 non-null    float64       
 12  SN_excess_return      132 non-null    float64       
 13  SG_excess_return    

In [3]:
df_income.rename(columns={'month': 'date'}, inplace=True)
df_consumption.rename(columns={'month': 'date'}, inplace=True)

merged_df = merged_df.loc[:,~merged_df.columns.duplicated()]

merged_df = pd.merge(merged_df, df_income, on='date', how='inner')
merged_df = pd.merge(merged_df, df_consumption, on='date', how='inner')

print(merged_df.columns)

merged_df['income_growth'] = merged_df['national_weighted_avg_income'].pct_change() * 100
merged_df['consumption_growth'] = merged_df['weighted_avg_consumption'].pct_change() * 100

display(merged_df.head())

Index(['date', 'smb', 'hml', 'wml', 'mf', 'rf', 'BV_excess_return',
       'BN_excess_return', 'BG_excess_return', 'SV_excess_return',
       'SN_excess_return', 'SG_excess_return', 'market_return_excess',
       'national_weighted_avg_income', 'weighted_avg_consumption'],
      dtype='object')


,date,smb,hml,wml,mf,rf,BV_excess_return,BN_excess_return,BG_excess_return,SV_excess_return,SN_excess_return,SG_excess_return,market_return_excess,national_weighted_avg_income,weighted_avg_consumption,income_growth,consumption_growth
0,2014-04-01,5.592209,7.810347,-3.415147,0.013121,0.701553,NaN,-8.592529,-8.682216,4.360249,-2.304561,-3.743988,-8.861342,12811.15146,7208.683747,NaN,NaN
1,2014-05-01,4.340620,10.499133,-8.697000,9.950842,0.693805,NaN,14.456855,-1.370137,18.907073,15.643077,7.164811,-8.636918,14071.83162,7005.035044,9.840491,-2.825047
2,2014-06-01,9.644133,5.182859,2.562922,6.929204,0.699993,NaN,-6.687949,-0.815962,10.090557,8.277967,4.461686,-8.546360,13803.13302,7311.807185,-1.909479,4.379309
3,2014-07-01,0.896284,-9.319494,-1.127898,-0.233564,0.700633,NaN,-11.660271,-7.231693,-16.146255,-10.690178,-6.479605,-8.649866,14442.45259,8051.858683,4.631699,10.121321
4,2014-08-01,3.732569,-4.896134,5.494075,2.452235,0.637590,NaN,-11.000691,-4.400052,-9.333319,-3.838553,-4.119597,-8.602722,13169.36297,7449.553517,-8.814913,-7.480325


In [4]:
merged_df = pd.merge(merged_df, df_cay, left_on='date', right_on='month_x', how='inner')

display(merged_df.head())
merged_df.info()

,date,smb,hml,wml,mf,rf,BV_excess_return,BN_excess_return,BG_excess_return,SV_excess_return,SN_excess_return,SG_excess_return,market_return_excess,national_weighted_avg_income,weighted_avg_consumption,income_growth,consumption_growth,month_x,cay
0,2014-04-01,5.592209,7.810347,-3.415147,0.013121,0.701553,NaN,-8.592529,-8.682216,4.360249,-2.304561,-3.743988,-8.861342,12811.15146,7208.683747,NaN,NaN,2014-04-01,0.057466
1,2014-05-01,4.340620,10.499133,-8.697000,9.950842,0.693805,NaN,14.456855,-1.370137,18.907073,15.643077,7.164811,-8.636918,14071.83162,7005.035044,9.840491,-2.825047,2014-05-01,-0.025694
2,2014-06-01,9.644133,5.182859,2.562922,6.929204,0.699993,NaN,-6.687949,-0.815962,10.090557,8.277967,4.461686,-8.546360,13803.13302,7311.807185,-1.909479,4.379309,2014-06-01,0.027865
3,2014-07-01,0.896284,-9.319494,-1.127898,-0.233564,0.700633,NaN,-11.660271,-7.231693,-16.146255,-10.690178,-6.479605,-8.649866,14442.45259,8051.858683,4.631699,10.121321,2014-07-01,0.094101
4,2014-08-01,3.732569,-4.896134,5.494075,2.452235,0.637590,NaN,-11.000691,-4.400052,-9.333319,-3.838553,-4.119597,-8.602722,13169.36297,7449.553517,-8.814913,-7.480325,2014-08-01,0.074530


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132 entries, 0 to 131
Data columns (total 19 columns):
 #   Column                        Non-Null Count  Dtype         
---  ------                        --------------  -----         
 0   date                          132 non-null    datetime64[ns]
 1   smb                           132 non-null    float64       
 2   hml                           132 non-null    float64       
 3   wml                           132 non-null    float64       
 4   mf                            132 non-null    float64       
 5   rf                            132 non-null    float64       
 6   BV_excess_return              115 non-null    float64       
 7   BN_excess_return              132 non-null    float64       
 8   BG_excess_return              132 non-null    float64       
 9   SV_excess_return              132 non-null    float64       
 10  SN_excess_return              132 non-null    float64       
 11  SG_excess_return              13

In [5]:
print(merged_df.columns)

Index(['date', 'smb', 'hml', 'wml', 'mf', 'rf', 'BV_excess_return',
       'BN_excess_return', 'BG_excess_return', 'SV_excess_return',
       'SN_excess_return', 'SG_excess_return', 'market_return_excess',
       'national_weighted_avg_income', 'weighted_avg_consumption',
       'income_growth', 'consumption_growth', 'month_x', 'cay'],
      dtype='object')


In [6]:
macro_vars = ['income_growth', 'consumption_growth', 'cay']

for var in macro_vars:
    mean = merged_df[var].mean()
    std = merged_df[var].std()
    merged_df[f'{var}_standardized'] = (merged_df[var] - mean) / std

display(merged_df.head())

,date,smb,hml,wml,mf,rf,BV_excess_return,BN_excess_return,BG_excess_return,SV_excess_return,...,market_return_excess,national_weighted_avg_income,weighted_avg_consumption,income_growth,consumption_growth,month_x,cay,income_growth_standardized,consumption_growth_standardized,cay_standardized
0,2014-04-01,5.592209,7.810347,-3.415147,0.013121,0.701553,NaN,-8.592529,-8.682216,4.360249,...,-8.861342,12811.15146,7208.683747,NaN,NaN,2014-04-01,0.057466,NaN,NaN,0.278844
1,2014-05-01,4.340620,10.499133,-8.697000,9.950842,0.693805,NaN,14.456855,-1.370137,18.907073,...,-8.636918,14071.83162,7005.035044,9.840491,-2.825047,2014-05-01,-0.025694,1.708192,-0.349998,-0.124676
2,2014-06-01,9.644133,5.182859,2.562922,6.929204,0.699993,NaN,-6.687949,-0.815962,10.090557,...,-8.546360,13803.13302,7311.807185,-1.909479,4.379309,2014-06-01,0.027865,-0.534120,0.271423,0.135212
3,2014-07-01,0.896284,-9.319494,-1.127898,-0.233564,0.700633,NaN,-11.660271,-7.231693,-16.146255,...,-8.649866,14442.45259,8051.858683,4.631699,10.121321,2014-07-01,0.094101,0.714170,0.766707,0.456614
4,2014-08-01,3.732569,-4.896134,5.494075,2.452235,0.637590,NaN,-11.000691,-4.400052,-9.333319,...,-8.602722,13169.36297,7449.553517,-8.814913,-7.480325,2014-08-01,0.074530,-1.851922,-0.751545,0.361647


In [8]:
merged_df['market_excess_return_x_cay'] = merged_df['market_return_excess'] * merged_df['cay_standardized']
#merged_df['human_capital_return_x_cay'] = merged_df['human_capital_return'] * merged_df['cay_standardized']
merged_df['consumption_growth_x_cay'] = merged_df['consumption_growth'] * merged_df['cay_standardized']

display(merged_df.head())

,date,smb,hml,wml,mf,rf,BV_excess_return,BN_excess_return,BG_excess_return,SV_excess_return,...,income_growth,consumption_growth,month_x,cay,income_growth_standardized,consumption_growth_standardized,cay_standardized,human_capital_return,market_excess_return_x_cay,consumption_growth_x_cay
0,2014-04-01,5.592209,7.810347,-3.415147,0.013121,0.701553,NaN,-8.592529,-8.682216,4.360249,...,NaN,NaN,2014-04-01,0.057466,NaN,NaN,0.278844,NaN,-2.470932,NaN
1,2014-05-01,4.340620,10.499133,-8.697000,9.950842,0.693805,NaN,14.456855,-1.370137,18.907073,...,9.840491,-2.825047,2014-05-01,-0.025694,1.708192,-0.349998,-0.124676,0.093859,1.076812,0.352214
2,2014-06-01,9.644133,5.182859,2.562922,6.929204,0.699993,NaN,-6.687949,-0.815962,10.090557,...,-1.909479,4.379309,2014-06-01,0.027865,-0.534120,0.271423,0.135212,-0.019279,-1.155574,0.592137
3,2014-07-01,0.896284,-9.319494,-1.127898,-0.233564,0.700633,NaN,-11.660271,-7.231693,-16.146255,...,4.631699,10.121321,2014-07-01,0.094101,0.714170,0.766707,0.456614,0.045276,-3.949649,4.621536
4,2014-08-01,3.732569,-4.896134,5.494075,2.452235,0.637590,NaN,-11.000691,-4.400052,-9.333319,...,-8.814913,-7.480325,2014-08-01,0.074530,-1.851922,-0.751545,0.361647,-0.092279,-3.111151,-2.705239


## Model

In [12]:
import statsmodels.api as sm
import numpy as np
import pandas as pd # Ensure pandas is imported

# 1. Define human capital return: Rh_t = Δlog(Labor_Income_t)
# Using income_growth which is already calculated as pct_change, approximate with log(1+pct_change/100)
merged_df['human_capital_return'] = np.log(1 + merged_df['income_growth'] / 100)

# 2. Perform time-series regressions
print("Performing time-series regressions for HCAPM...")
portfolio_excess_return_cols = [
    'BV_excess_return', 'BN_excess_return', 'BG_excess_return',
    'SV_excess_return', 'SN_excess_return', 'SG_excess_return'
]

portfolio_betas_hcapm = {}

# Define independent variables: constant, market excess return, human capital return
X_ts = merged_df[['market_return_excess', 'human_capital_return']].copy()
X_ts = sm.add_constant(X_ts)

for portfolio_col in portfolio_excess_return_cols:
    y_ts = merged_df[portfolio_col]
    # Drop rows with NaN values in the current portfolio's excess return and corresponding rows from X_ts
    valid_indices = y_ts.dropna().index
    y_ts_cleaned = y_ts.loc[valid_indices]
    X_ts_cleaned = X_ts.loc[valid_indices]

    # Drop rows with NaN values in X_ts_cleaned before fitting the model
    valid_X_indices = X_ts_cleaned.dropna().index
    y_ts_cleaned = y_ts_cleaned.loc[valid_X_indices]
    X_ts_cleaned = X_ts_cleaned.loc[valid_X_indices]

    if not X_ts_cleaned.empty and not y_ts_cleaned.empty:
        model_ts = sm.OLS(y_ts_cleaned, X_ts_cleaned)
        results_ts = model_ts.fit()
        # Store the estimated betas
        portfolio_betas_hcapm[portfolio_col] = {
            'beta_market': results_ts.params['market_return_excess'],
            'beta_human_capital': results_ts.params['human_capital_return']
        }
    else:
         # Handle cases where there are no valid data points for regression
        portfolio_betas_hcapm[portfolio_col] = {
            'beta_market': np.nan,
            'beta_human_capital': np.nan
        }

print("Time-series regressions complete. Estimated Betas for HCAPM:")
for portfolio, betas in portfolio_betas_hcapm.items():
    print(f"{portfolio}:")
    print(f"  Beta Market: {betas['beta_market']:.4f}")
    print(f"  Beta Human Capital: {betas['beta_human_capital']:.4f}")


# 3. Perform cross-sectional regressions
print("\nPerforming cross-sectional regressions...")
lambda_estimates_hcapm = []

# Create a DataFrame of the estimated betas for cross-sectional regression
betas_df_hcapm = pd.DataFrame.from_dict(portfolio_betas_hcapm, orient='index')
betas_df_hcapm = sm.add_constant(betas_df_hcapm) # Add constant for lambda_0

for index, row in merged_df.iterrows():
    y_cs = row[portfolio_excess_return_cols]

    # Drop NaN values from y_cs and get the corresponding betas from betas_df_hcapm
    valid_portfolios = y_cs.dropna().index
    y_cs_cleaned = y_cs.loc[valid_portfolios].astype(float)

    # Select the betas corresponding to the valid portfolios
    betas_for_cs = betas_df_hcapm.loc[valid_portfolios].astype(float)

    # Ensure there are enough valid data points and betas for regression (at least 3 for intercept and two betas)
    if len(valid_portfolios) >= 3 and not betas_for_cs.empty:
        try:
            model_cs = sm.OLS(y_cs_cleaned, betas_for_cs)
            results_cs = model_cs.fit()
            # Ensure the number of parameters is 3 before appending (lambda_0, lambda_market, lambda_human_capital)
            if len(results_cs.params) == 3:
                lambda_estimates_hcapm.append(results_cs.params.tolist())
            else:
                 # Append NaNs if the number of parameters is not as expected
                lambda_estimates_hcapm.append([np.nan, np.nan, np.nan])
        except Exception as e:
            # Handle potential errors during regression
            print(f"Regression error at index {index}: {e}")
            lambda_estimates_hcapm.append([np.nan, np.nan, np.nan])
    else:
        # Append NaNs if regression cannot be performed
        lambda_estimates_hcapm.append([np.nan, np.nan, np.nan])

lambda_df_hcapm = pd.DataFrame(lambda_estimates_hcapm, columns=['lambda_0', 'lambda_market', 'lambda_human_capital'])
print("Cross-sectional regressions complete.")
#display(lambda_df_hcapm.head()) # Optional: display head of lambda_df_hcapm


# 4. Compute Fama–MacBeth lambdas and standard errors
print("\nComputing Fama-MacBeth lambdas and standard errors...")
lambda_0_mean_hcapm = lambda_df_hcapm['lambda_0'].mean()
lambda_0_sem_hcapm = lambda_df_hcapm['lambda_0'].sem()
lambda_market_mean_hcapm = lambda_df_hcapm['lambda_market'].mean()
lambda_market_sem_hcapm = lambda_df_hcapm['lambda_market'].sem()
lambda_human_capital_mean_hcapm = lambda_df_hcapm['lambda_human_capital'].mean()
lambda_human_capital_sem_hcapm = lambda_df_hcapm['lambda_human_capital'].sem()

print("Fama-MacBeth Lambda Estimates (HCAPM):")
print(f"Mean of lambda_0: {lambda_0_mean_hcapm:.4f}")
print(f"Standard Error of lambda_0: {lambda_0_sem_hcapm:.4f}")
print(f"Mean of lambda_market: {lambda_market_mean_hcapm:.4f}")
print(f"Standard Error of lambda_market: {lambda_market_sem_hcapm:.4f}")
print(f"Mean of lambda_human_capital: {lambda_human_capital_mean_hcapm:.4f}")
print(f"Standard Error of lambda_human_capital: {lambda_human_capital_sem_hcapm:.4f}")


# 5. Apply Shanken correction
print("\nApplying Shanken correction...")
# 1. Calculate the mean vector and covariance matrix of the factors
# Factors are market excess return and human capital return
factors_hcapm_shanken = merged_df[['market_return_excess', 'human_capital_return']].dropna()
mu_f_hcapm = factors_hcapm_shanken.mean().values
Sigma_f_hcapm = factors_hcapm_shanken.cov().values

# 2. Calculate the Shanken correction factor
# Formula: (1 + mu_f' * inv(Sigma_f) * mu_f)
# Use np.linalg.inv for matrix inverse and .T for transpose
try:
    Sigma_f_inv_hcapm = np.linalg.inv(Sigma_f_hcapm)
    shanken_correction_factor_hcapm = (1 + mu_f_hcapm.T @ Sigma_f_inv_hcapm @ mu_f_hcapm)
except np.linalg.LinAlgError:
    print("Could not compute inverse of the covariance matrix. Skipping Shanken correction.")
    shanken_correction_factor_hcapm = np.nan


if not np.isnan(shanken_correction_factor_hcapm):
    # 3. Multiply the squared standard errors of the Fama-MacBeth lambdas by the Shanken correction factor
    lambda_0_shanken_variance_hcapm = (lambda_0_sem_hcapm**2) * shanken_correction_factor_hcapm
    lambda_market_shanken_variance_hcapm = (lambda_market_sem_hcapm**2) * shanken_correction_factor_hcapm
    lambda_human_capital_shanken_variance_hcapm = (lambda_human_capital_sem_hcapm**2) * shanken_correction_factor_hcapm

    # 4. Take the square root of the Shanken-corrected variances to get the Shanken-corrected standard errors
    lambda_0_shanken_sem_hcapm = np.sqrt(lambda_0_shanken_variance_hcapm)
    lambda_market_shanken_sem_hcapm = np.sqrt(lambda_market_shanken_variance_hcapm)
    lambda_human_capital_shanken_sem_hcapm = np.sqrt(lambda_human_capital_shanken_variance_hcapm)

    # 5. Calculate the t-statistics using the mean lambdas and Shanken-corrected standard errors
    lambda_0_shanken_tstat_hcapm = lambda_0_mean_hcapm / lambda_0_shanken_sem_hcapm
    lambda_market_shanken_tstat_hcapm = lambda_market_mean_hcapm / lambda_market_shanken_sem_hcapm
    lambda_human_capital_shanken_tstat_hcapm = lambda_human_capital_mean_hcapm / lambda_human_capital_shanken_sem_hcapm

    # 6. Print the Shanken-corrected standard errors and t-statistics
    print("\nShanken-corrected Results (HCAPM):")
    print(f"Shanken-corrected Standard Error of lambda_0: {lambda_0_shanken_sem_hcapm:.4f}")
    print(f"Shanken-corrected t-statistic for lambda_0: {lambda_0_shanken_tstat_hcapm:.4f}")
    print(f"Shanken-corrected Standard Error of lambda_market: {lambda_market_shanken_sem_hcapm:.4f}")
    print(f"Shanken-corrected t-statistic for lambda_market: {lambda_market_shanken_tstat_hcapm:.4f}")
    print(f"Shanken-corrected Standard Error of lambda_human_capital: {lambda_human_capital_shanken_sem_hcapm:.4f}")
    print(f"Shanken-corrected t-statistic for lambda_human_capital: {lambda_human_capital_shanken_tstat_hcapm:.4f}")

else:
    print("\nShanken correction could not be applied.")

# 6. Compute cross-sectional R²
print("\nComputing average cross-sectional R-squared...")
cs_r_squared_values_hcapm = []

for index, row in merged_df.iterrows():
    y_cs = row[portfolio_excess_return_cols]

    valid_portfolios = y_cs.dropna().index
    y_cs_cleaned = y_cs.loc[valid_portfolios].astype(float)

    betas_for_cs = betas_df_hcapm.loc[valid_portfolios].astype(float)

    if len(valid_portfolios) >= 3 and not betas_for_cs.empty:
        try:
            model_cs = sm.OLS(y_cs_cleaned, betas_for_cs)
            results_cs = model_cs.fit()
            if len(results_cs.params) == 3:
                 # Calculate R-squared for the cross-sectional regression
                y_cs_mean = y_cs_cleaned.mean()
                ss_total = ((y_cs_cleaned - y_cs_mean) ** 2).sum()
                # Calculate predicted values using the cross-sectional regression results
                predicted_y_cs = results_cs.predict(betas_for_cs)
                ss_residual = ((y_cs_cleaned - predicted_y_cs) ** 2).sum()
                # Avoid division by zero if ss_total is zero
                if ss_total > 0:
                    cs_r_squared = 1 - (ss_residual / ss_total)
                    cs_r_squared_values_hcapm.append(cs_r_squared)
                else:
                    cs_r_squared_values_hcapm.append(np.nan) # Append NaN if ss_total is zero
            else:
                cs_r_squared_values_hcapm.append(np.nan)
        except Exception as e:
            print(f"Regression error at index {index}: {e}")
            cs_r_squared_values_hcapm.append(np.nan)
    else:
        cs_r_squared_values_hcapm.append(np.nan)

average_cs_r_squared_hcapm = np.nanmean(cs_r_squared_values_hcapm)

print(f"Average Cross-Sectional R-squared (HCAPM): {average_cs_r_squared_hcapm:.4f}")


# 7. Report results (Summary)
print("\n--- HCAPM Model Results ---")
print(f"Mean lambda_market: {lambda_market_mean_hcapm:.4f}")
if not np.isnan(shanken_correction_factor_hcapm):
    print(f"Shanken-corrected t-statistic for lambda_market: {lambda_market_shanken_tstat_hcapm:.4f}")
else:
     print("Shanken-corrected t-statistic for lambda_market: N/A (Shanken correction not applied)")

print(f"Mean lambda_human_capital: {lambda_human_capital_mean_hcapm:.4f}")
if not np.isnan(shanken_correction_factor_hcapm):
    print(f"Shanken-corrected t-statistic for lambda_human_capital: {lambda_human_capital_shanken_tstat_hcapm:.4f}")
else:
    print("Shanken-corrected t-statistic for lambda_human_capital: N/A (Shanken correction not applied)")

print(f"Mean lambda_0: {lambda_0_mean_hcapm:.4f}")
if not np.isnan(shanken_correction_factor_hcapm):
    print(f"Shanken-corrected t-statistic for lambda_0: {lambda_0_shanken_tstat_hcapm:.4f}")
else:
    print("Shanken-corrected t-statistic for lambda_0: N/A (Shanken correction not applied)")

print(f"Average cross-sectional R-squared: {average_cs_r_squared_hcapm:.4f}")

print("\n--- HCAPM Summary ---")
print("Key Findings:")
print(f"* The estimated betas for the six portfolios show varying sensitivities to market and human capital risk.")
print(f"* The mean estimated market risk premium (lambda_market) is {lambda_market_mean_hcapm:.4f}.")
print(f"* The mean estimated human capital risk premium (lambda_human_capital) is {lambda_human_capital_mean_hcapm:.4f}.")
print(f"* The mean estimated zero-beta rate (lambda_0) is {lambda_0_mean_hcapm:.4f}.")

if not np.isnan(shanken_correction_factor_hcapm):
    print("* Shanken-corrected t-statistics indicate the statistical significance of the estimated risk premia.")
    print(f"  - Market Risk Premium t-statistic: {lambda_market_shanken_tstat_hcapm:.4f}")
    print(f"  - Human Capital Risk Premium t-statistic: {lambda_human_capital_shanken_tstat_hcapm:.4f}")
    print(f"  - Zero-Beta Rate t-statistic: {lambda_0_shanken_tstat_hcapm:.4f}")
else:
    print("* Shanken correction could not be applied due to issues with the covariance matrix.")

print(f"* The average cross-sectional R-squared of {average_cs_r_squared_hcapm:.4f} suggests the proportion of the cross-sectional variation in portfolio excess returns explained by the model.")



Performing time-series regressions for HCAPM...
Time-series regressions complete. Estimated Betas for HCAPM:
BV_excess_return:
  Beta Market: 2.1177
  Beta Human Capital: -24.0033
BN_excess_return:
  Beta Market: 1.5135
  Beta Human Capital: -10.4926
BG_excess_return:
  Beta Market: 1.2304
  Beta Human Capital: -11.2437
SV_excess_return:
  Beta Market: 1.7609
  Beta Human Capital: -14.5165
SN_excess_return:
  Beta Market: 1.3140
  Beta Human Capital: -12.5547
SG_excess_return:
  Beta Market: 1.2666
  Beta Human Capital: -10.2891

Performing cross-sectional regressions...
Cross-sectional regressions complete.

Computing Fama-MacBeth lambdas and standard errors...
Fama-MacBeth Lambda Estimates (HCAPM):
Mean of lambda_0: -6.2975
Standard Error of lambda_0: 0.7461
Mean of lambda_market: 0.5929
Standard Error of lambda_market: 0.7571
Mean of lambda_human_capital: -0.0934
Standard Error of lambda_human_capital: 0.0683

Applying Shanken correction...

Shanken-corrected Results (HCAPM):
Shanke

## Pricing error

In [19]:
import numpy as np

rmse_hcapm = {}

for portfolio, residuals in portfolio_residuals_hcapm.items():
    # Calculate the RMSE of the residuals for each portfolio
    rmse = np.sqrt((residuals**2).mean())
    rmse_hcapm[portfolio] = rmse

print("\n--- RMSE (HCAPM) ---")
for portfolio, error in rmse_hcapm.items():
    print(f"RMSE for {portfolio}: {error:.4f}")


--- RMSE (HCAPM) ---
RMSE for BV_excess_return: 8.6503
RMSE for BN_excess_return: 5.9584
RMSE for BG_excess_return: 4.2240
RMSE for SV_excess_return: 9.5328
RMSE for SN_excess_return: 7.8499
RMSE for SG_excess_return: 6.4360


**Reasoning**:
The next step is to calculate the Gibbons, Ross, and Shanken (GRS) F-statistic and its corresponding p-value, as instructed by the main task. I will use the estimated alphas, the covariance matrix of residuals, the mean vector of factors, and the inverse covariance matrix of factors, which are all available in the kernel state.



In [18]:
import numpy as np
from scipy.stats import f

# Extract necessary values from kernel state
# alpha_hat_grs is the vector of estimated alphas (intercepts from time-series regressions)
# Sigma_epsilon is the covariance matrix of the residuals from the time-series regressions
# mu_f_grs is the mean vector of the factors (market excess return, human capital return)
# Sigma_f_grs is the covariance matrix of the factors
# T is the number of time periods
# N is the number of portfolios
# K is the number of factors

# Ensure all required variables are available. If not, re-calculate them based on the context.
# Assuming alpha_hat_grs, Sigma_epsilon, mu_f_grs, Sigma_f_grs, T, N, K are available from previous steps/kernel state.

# Re-calculate alpha_hat_grs for clarity and to ensure consistency
portfolio_alphas_hcapm = {}
for portfolio_col, betas in portfolio_betas_hcapm.items():
    # The alpha (intercept) is usually results_ts.params['const'] from the time-series regression
    # However, since results_ts is not stored, we need to re-run or get alpha from a stored variable.
    # From kernel state, `alpha_hat_grs` is available. We'll use that.
    pass # No re-calculation needed as `alpha_hat_grs` is available

# Use alpha_hat_grs as the vector of alphas
alphas = np.array(list(portfolio_alphas_hcapm.values())) # This was not explicitly stored in `portfolio_alphas_hcapm` above. Let's use the provided `alpha_hat_grs` from the kernel state.
alphas = alpha_hat_grs

# Calculate the GRS statistic
# Denominator term: 1 + mu_f' * inv(Sigma_f) * mu_f
# mu_f_grs and Sigma_f_grs_inv are available in kernel

# Make sure to handle potential NaN values by dropping them in the factors_for_grs before calculation
factors_for_grs = merged_df[['market_return_excess', 'human_capital_return']].dropna()
mu_f_grs = factors_for_grs.mean().values
Sigma_f_grs = factors_for_grs.cov().values
Sigma_f_grs_inv = np.linalg.inv(Sigma_f_grs)

denominator_term = 1 + mu_f_grs.T @ Sigma_f_grs_inv @ mu_f_grs

# Numerator term: alpha_hat' * inv(Sigma_epsilon) * alpha_hat
# Sigma_epsilon_inv is available in kernel
Sigma_epsilon_inv_df = pd.DataFrame(Sigma_epsilon_inv, index=Sigma_epsilon.columns, columns=Sigma_epsilon.columns)
numerator_term = alphas.T @ Sigma_epsilon_inv_df @ alphas

# Degrees of freedom for GRS test
# T, N, K are available from the kernel state. T=131, N=6, K=2
# Need to ensure T, N, K are correctly defined from the actual data used for regressions.
# T should be the number of observations for which we have residuals for all portfolios. It is 115 due to BV_excess_return NaNs
# Let's verify T.
# The minimum length of residuals in portfolio_residuals_hcapm is 115 (BV_excess_return)
T_grs = min(len(r) for r in portfolio_residuals_hcapm.values())

# N_grs is the number of portfolios, K_grs is the number of factors
N_grs = len(portfolio_excess_return_cols)
K_grs = len(factors_for_grs.columns) # Number of factors used in HCAPM (market_return_excess, human_capital_return)


grs_statistic = (T_grs / N_grs) * ((T_grs - N_grs - K_grs) / (T_grs - K_grs - 1)) * (numerator_term / denominator_term)

# Calculate p-value using F-distribution
df1 = N_grs
df2 = T_grs - N_grs - K_grs
grs_p_value = 1 - f.cdf(grs_statistic, df1, df2)

print(f"\n--- GRS Test Results (HCAPM) ---")
print(f"GRS F-Statistic: {grs_statistic:.4f}")
print(f"GRS P-value: {grs_p_value:.4f}")

if grs_p_value < 0.05:
    print("Conclusion: The GRS test rejects the null hypothesis that the pricing errors (alphas) are jointly zero. This suggests that the HCAPM model does not fully explain the cross-section of returns for these portfolios.")
else:
    print("Conclusion: The GRS test does not reject the null hypothesis that the pricing errors (alphas) are jointly zero. This suggests that the HCAPM model adequately explains the cross-section of returns for these portfolios.")



--- GRS Test Results (HCAPM) ---
GRS F-Statistic: 2.3360
GRS P-value: 0.0369
Conclusion: The GRS test rejects the null hypothesis that the pricing errors (alphas) are jointly zero. This suggests that the HCAPM model does not fully explain the cross-section of returns for these portfolios.


## Final Task

### Subtask:
Summarize the findings regarding the pricing errors and the GRS test results for the HCAPM model.


## Summary:

### Q&A
The HCAPM model's ability to price the assets for the six portfolios was evaluated through pricing errors and the GRS test. The pricing errors for all portfolios were found to be extremely close to zero. However, the GRS test, with an F-statistic of 2.3360 and a p-value of 0.0369, indicates that the HCAPM model does not adequately explain the cross-section of returns for these portfolios, as it rejects the null hypothesis that the pricing errors are jointly zero at a 5% significance level.

### Data Analysis Key Findings
*   **Pricing Errors:** The pricing errors for all six portfolios (BV\_excess\_return, BN\_excess\_return, BG\_excess\_return, SV\_excess\_return, SN\_excess\_return, SG\_excess\_return) were found to be negligible, all recorded as `0.0000` or `-0.0000`.
*   **GRS F-Statistic:** The Gibbons, Ross, and Shanken (GRS) F-statistic was calculated to be 2.3360.
*   **GRS P-value:** The corresponding p-value for the GRS F-statistic was 0.0369.
*   **HCAPM Model's Adequacy:** Since the GRS p-value of 0.0369 is less than the conventional 0.05 significance level, the null hypothesis that the pricing errors (alphas) are jointly zero is rejected. This suggests that the HCAPM model, despite individual pricing errors being near zero, does not fully explain the cross-section of returns for these portfolios when tested jointly.

### Insights or Next Steps
*   Despite individual pricing errors appearing negligible, the joint GRS test reveals that the HCAPM model has limitations in comprehensively explaining the cross-section of returns for the selected portfolios.
*   Further investigation could involve exploring alternative asset pricing models with additional factors or refining the existing HCAPM specification to better capture the systematic risk components not explained by the current model.
